automotive/odcantools/libodcantools/src/CANMessage.cpp

```diff
@@ -0,0 +1,365 @@
+        }
+
+        CANMessage::CANMessage(const CANMessage& canmessage) :
+            GenericCANMessage(canmessage),
+            m_payload(canmessage.m_payload),
+            m_signals(canmessage.m_signals)
+            {
+        }
+        
+        void CANMessage::setFromUINT64(const uint64_t payload, const uint8_t length) {
+            // superclass methods
+            setData(payload);
+            setLength(length);
+            
+            uint64_t data=payload;
+            uint8_t mask=0xFF;
+            
+            for(uint8_t i=0;i<length;++i) {
+                m_payload.push_back(data & mask);
+                data=data>>8;
+            }
+            
+            // invert payload due to the way it is represented in log files or traces
+            m_payload=invertPayloadVector(m_payload);
+        }
+        
+        vector<uint8_t> CANMessage::invertPayloadVector(vector<uint8_t> payload) {
+            vector<uint8_t> temp;
+            for(int8_t i=payload.size()-1;i>=0;--i) {
+                temp.push_back(payload.at(i));
+            }
+            return temp;
+        }
+        
+        uint64_t CANMessage::getPayloadToUINT64() {
+            // invert payload due to the way it is represented in log files or traces
+            vector<uint8_t> temp=m_payload;
+            temp=invertPayloadVector(temp);
+            
+            uint64_t data=0;
+            
+            for(int8_t i=temp.size()-1;i>=0;--i) {
+                data=data<<8;
+                data = data | temp.at(i);
+            }
+            
+            return data;
+        }
+        
+        automotive::GenericCANMessage CANMessage::getGenericCANMessage() {
+            automotive::GenericCANMessage gcm;
+            
+            // use superclass getters
+            gcm.setIdentifier(getIdentifier());
+            gcm.setLength(getLength());
+            gcm.setData(getData());
+            
+            return gcm;
+        }
+        
+        bool CANMessage::addSignal(const uint16_t key, const CANSignal signal) {
+            std::pair<std::map<const uint16_t, const CANSignal>::iterator,bool> result;
+            result = m_signals.insert ( std::pair<const uint16_t, const CANSignal>(key, signal) );
+            return result.second;
+        }
+        
+        void CANMessage::advanceByteMask(uint8_t& byteMask) {
+            // both Intel Standard and Motorola Forward MSB number bits in a byte from the msb to the lsb (7 6 5 4 3 2 1 0)
+            byteMask=byteMask<<1;
+        }
+
+        void CANMessage::resetByteMask(uint8_t& byteMask) {
+            // move the mask back to the extreme right (bit 0 in a byte)
+            byteMask=0x01;
+        }
+
+        void CANMessage::advanceByteNumberInCorrectEndianness(int8_t& byteNumber, const string endianness) {
+            // if the signal is little endian, increase the byte number
+            if(endianness.compare("little")==0) {
```

Comment:
The `string::compare`might be expensive --> would there be a better implementation?